mlflow/models/evaluation/base.py

```diff
@@ -456,11 +522,141 @@ def _start_run_or_reuse_active_run(run_id):
+                "they will all be used in evaluation if they support the specified model type. "
+                "If you want to evaluate with one evaluator, specify the `evaluator` argument "
+                "and optionally specify the `evaluator_config` argument."
+            )
+        if evaluator_config is not None:
+            conf_dict_value_error = ValueError(
+                "If `evaluators` argument is None, all available evaluators will be used. "
+                "If only the default evaluator is available, the `evaluator_config` argument is "
+                "interpreted as the config dictionary for the default evaluator. Otherwise, the "
+                "`evaluator_config` argument must be a dictionary mapping each evaluator's name "
+                "to its own evaluator config dictionary."
+            )
+            if evaluator_name_list == ["default"]:
+                if not isinstance(evaluator_config, dict):
+                    raise conf_dict_value_error
+                elif "default" not in evaluator_config:
+                    evaluator_name_to_conf_map = {"default": evaluator_config}
+                else:
+                    evaluator_name_to_conf_map = evaluator_config
+            else:
+                if not check_nesting_config_dict(evaluator_name_list, evaluator_config):
+                    raise conf_dict_value_error
+                evaluator_name_to_conf_map = evaluator_config
+        else:
+            evaluator_name_to_conf_map = {}
+    elif isinstance(evaluators, str):
+        if not (evaluator_config is None or isinstance(evaluator_config, dict)):
+            raise ValueError(
+                "If `evaluators` argument is the name of an evaluator, evaluator_config must be "
+                "None or a dict containing config items for the evaluator."
+            )
+        evaluator_name_list = [evaluators]
+        evaluator_name_to_conf_map = {evaluators: evaluator_config}
+    elif isinstance(evaluators, list):
+        if evaluator_config is not None:
+            if not check_nesting_config_dict(evaluators, evaluator_config):
+                raise ValueError(
+                    "If `evaluators` argument is an evaluator name list, evaluator_config "
+                    "must be a dict contains mapping from evaluator name to individual "
+                    "evaluator config dict."
+                )
+        # Use `OrderedDict.fromkeys` to deduplicate elements but keep elements order.
+        evaluator_name_list = list(OrderedDict.fromkeys(evaluators))
+        evaluator_name_to_conf_map = evaluator_config or {}
+    else:
+        raise ValueError(
+            "`evaluators` argument must be None, an evaluator name string, or a list of "
+            "evaluator names."
+        )
+
+    return evaluator_name_list, evaluator_name_to_conf_map
+
+
+_last_failed_evaluator = None
+
+
+def _get_last_failed_evaluator():
```

Comment:
where do we use this function?

Reply:
In usage log, we want to log error raised from which evaluator (TODO), for debug purpose. so I add this.

Reply:
why do we only store the last failed evaluator?

Reply:
The evaluate API will raise error when it hit the first failed evaluator. (The API designed to be so.) When any evaluator fail `evaluate` raise error, otherwise return all merged success results